crates/neon/src/types_impl/extract/buffer.rs

```diff
@@ -0,0 +1,240 @@
+            Err(_) => return Ok(Err(Self::Error::new())),
+        };
+
+        Ok(Ok(ArrayBuffer(B::from(v.as_slice(cx)))))
+    }
+}
+
+impl<'cx, B> TryIntoJs<'cx> for ArrayBuffer<B>
+where
+    B: AsRef<[u8]>,
+{
+    type Value = JsArrayBuffer;
+
+    fn try_into_js(self, cx: &mut Cx<'cx>) -> JsResult<'cx, Self::Value> {
+        JsArrayBuffer::from_slice(cx, self.0.as_ref())
+    }
+}
+
+impl<B> private::Sealed for ArrayBuffer<B> {}
+
+/// Wrapper for converting between bytes and [`JsBuffer`](JsBuffer)
+pub struct Buffer<B>(pub B);
+
+impl<'cx, B> TryFromJs<'cx> for Buffer<B>
+where
+    for<'b> B: From<&'b [u8]>,
+{
+    type Error = TypeExpected<JsBuffer>;
+
+    fn try_from_js(
+        cx: &mut Cx<'cx>,
+        v: Handle<'cx, JsValue>,
+    ) -> NeonResult<Result<Self, Self::Error>> {
+        let v = match v.downcast::<JsBuffer, _>(cx) {
+            Ok(v) => v,
+            Err(_) => return Ok(Err(Self::Error::new())),
+        };
+
+        Ok(Ok(Buffer(B::from(v.as_slice(cx)))))
+    }
+}
+
+impl<'cx, B> TryIntoJs<'cx> for Buffer<B>
+where
+    B: AsRef<[u8]>,
+{
+    type Value = JsBuffer;
+
+    fn try_into_js(self, cx: &mut Cx<'cx>) -> JsResult<'cx, Self::Value> {
+        JsBuffer::from_slice(cx, self.0.as_ref())
+    }
+}
+
+impl<B> private::Sealed for Buffer<B> {}
+
+impl<'cx, T> TryIntoJs<'cx> for Vec<T>
+where
+    JsTypedArray<T>: Value,
+    T: Binary,
+{
+    type Value = JsTypedArray<T>;
+
+    fn try_into_js(self, cx: &mut Cx<'cx>) -> JsResult<'cx, Self::Value> {
+        JsTypedArray::from_slice(cx, self.as_slice())
+    }
+}
+
+impl<'cx, T> TryFromJs<'cx> for Vec<T>
+where
+    JsTypedArray<T>: Value,
+    T: Binary,
+{
+    type Error = TypeExpected<JsTypedArray<T>>;
+
+    fn try_from_js(
+        cx: &mut Cx<'cx>,
+        v: Handle<'cx, JsValue>,
+    ) -> NeonResult<Result<Self, Self::Error>> {
+        let v = match v.downcast::<JsTypedArray<T>, _>(cx) {
+            Ok(v) => v,
+            Err(_) => return Ok(Err(Self::Error::new())),
+        };
+
+        Ok(Ok(v.as_slice(cx).to_vec()))
+    }
+}
+
+impl<T> private::Sealed for Vec<T>
+where
+    JsTypedArray<T>: Value,
+    T: Binary,
+{
+}
+
+impl<'cx, T> TryIntoJs<'cx> for Box<[T]>
+where
+    JsTypedArray<T>: Value,
+    T: Binary,
+{
+    type Value = JsTypedArray<T>;
+
+    fn try_into_js(self, cx: &mut Cx<'cx>) -> JsResult<'cx, Self::Value> {
+        JsTypedArray::from_slice(cx, &self)
+    }
+}
+
+impl<T> private::Sealed for Box<[T]>
+where
+    JsTypedArray<T>: Value,
+    T: Binary,
+{
+}
+
+impl<'cx, T, const N: usize> TryIntoJs<'cx> for [T; N]
+where
+    JsTypedArray<T>: Value,
+    T: Binary,
+{
+    type Value = JsTypedArray<T>;
+
+    fn try_into_js(self, cx: &mut Cx<'cx>) -> JsResult<'cx, Self::Value> {
+        JsTypedArray::from_slice(cx, self.as_slice())
+    }
+}
+
+impl<T, const N: usize> private::Sealed for [T; N]
+where
+    JsTypedArray<T>: Value,
+    T: Binary,
+{
+}
+
+impl<'cx, T> TryIntoJs<'cx> for &[T]
```

Comment:
This may be a corner case but should we also think about impls for `std::cell:Ref` as well? It's very similar to `&` but comes up when you dynamically borrow from a standard cell, and if we don't implement it the orphan rule will prevent anyone else from being able to.

Reply:
Oh never mind, that's covered in the wrappers via `AsRef`.

Reply:
Yep! It will work with anything that is `AsRef<[T]>` for `TryIntoJs` and `for<'a> From<&'a [T]>` for `TryFromJs`. This covers `Ref` as well as other useful types like `bytes::Bytes`.

Keep in mind that in _many_ cases you wouldn't actually be able to use `Ref` directly because of the lifetime, but there are some niche scenarios where it would work.